src/hooks/useImmediateState.ts

```diff
@@ -0,0 +1,33 @@
+import { useRef, useEffect, useState, useCallback } from 'react';
+
+type SetStateAction<S> = S | ((prevState: S) => S);
+type SetState<S> = (action: SetStateAction<S>) => void;
+
+/** This is a drop-in replacement for useState, limited to object-based state. During initial mount it will mutably update the state, instead of scheduling a React update using setState */
+export const useImmediateState = <S extends {}>(init: S): [S, SetState<S>] => {
+  const isMounted = useRef(false);
+  const initialState = useRef<S>({ ...init });
+  const [state, setState] = useState<S>(initialState.current);
+
+  // This wraps setState and updates the state mutably on initial mount
+  // It also prevents setting the state when the component is unmounted
+  const updateState: SetState<S> = useCallback((action: SetStateAction<S>) => {
+    if (isMounted.current) {
+      setState(action);
+    } else if (typeof action === 'function') {
+      const update = (action as any)(initialState.current);
+      Object.assign(initialState.current, update);
```

Comment:
Alright, totally could be missing something, but don't we need to do the assignment to `initialState.current` here to ensure the mutable ref gets updated, i.e.

```
initialState.current = Object.assign(initialState.current, update)
```

Reply:
Nah, the first argument is being assigned to mutably 😀

Reply:
Ah right, I forgot that `Object.assign` mutates the original object 👍